File: src/intro_tf2/src/intro_tf2_listener.cpp
#include <ros/ros.h>
#include <tf2_ros/transform_listener.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>

class ListenerNode
{
public:
  ListenerNode()
    : source_frame_("base_link") //
    , target_frame_("map")
  { }

  bool configure()
  {
    tf2_buffer_.reset(new tf2_ros::Buffer);
    tf2_listener_.reset(new tf2_ros::TransformListener(*tf2_buffer_));
    return true;
  }

  void get_robot_pose(const ros::Time& stamp = ros::Time(0))
  {
    tf2::Transform tf; // useful if you want to work with your transform
    try
    {
      tf2::fromMsg( // convert geometry_msgs::Transform to tf2::Transform
        tf2_buffer_->lookupTransform(
          target_frame_,
          source_frame_,
          stamp).transform,
        tf);
    }
    catch (tf2::TransformException &ex)
    {
      ROS_WARN_THROTTLE(1, "%s", ex.what());
      return;
    }
    ROS_INFO_STREAM(tf.getOrigin().x() << " " <<
                    tf.getOrigin().y() << " " <<
                    tf.getOrigin().z());
  }

  void spin()
  {
    ros::Rate rate(2);
    while (ros::ok())
    {
      get_robot_pose();
      ros::spinOnce();
      rate.sleep();
    }
  }

private:
  boost::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
  boost::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
  std::string source_frame_, target_frame_;
};

int main(int argc, char** argv)
{
  ros::init(argc, argv, "tf2_listener_node");
  ListenerNode node;
  node.configure();
  node.spin();

  return EXIT_SUCCESS;
}

File: src/intro_tf2/src/intro_tf2_listener.py
#!/usr/bin/env python
import rospy

import tf
import tf2_ros

class ListenerNode(object):
    def __init__(self):
        self._source_frame = 'base_link'
        self._target_frame = 'map'

    def configure(self):
        self._tf2_buffer = tf2_ros.Buffer()
        self._tf2_listener = tf2_ros.TransformListener(self._tf2_buffer)
        return True;

    def get_robot_pose(self, stamp=rospy.Time(0)):
        try:
            tf = self._tf2_buffer.lookup_transform(
                self._target_frame,
                self._source_frame,
                stamp).transform
        except tf2_ros.TransformException as ex:
            rospy.logwarn_throttle(1, '%s' % ex)
            return

        rospy.loginfo("%s %s %s" % (tf.translation.x,
                                    tf.translation.y,
                                    tf.translation.z))

    def spin(self):
        rate = rospy.Rate(2)
        while not rospy.is_shutdown():
            self.get_robot_pose()
            rate.sleep()


if __name__ == '__main__':
    rospy.init_node('tf2_listener_node')
    node = ListenerNode()
    node.configure()
    node.spin()
